Update Dockerfile with command line parameters including bootstrapping option

JorritvandenBerg opened the pull request:
**What current issue(s) does this address, or what feature is it adding?**
Users having trouble to bootstrap a test chain. This is an update to the Dockerfile, which enables the use of prompt.py and bootstrap.py command line parameters in Docker and introducing an additional --bootstrap parameter. This parameter allows users to bootstrap prior to starting the prompt.

**How did you solve this problem?**
Creating a bash script (entrypoint.sh) for this procedure.

**How did you make sure your solution works?**
Tested various combinations of parameters and created a fallback for combinations that don't work together as per the comments in prompt.py/bootstrap.py. But please experiment with to iron out any edge cases.

**Are there any special changes in the code that we should be aware of?**
As an ENTRYPOINT is defined in the Docker file, one can not enter a bash shell anymore in the previous way. I have updated the example in the README and Dockerfile to explain how to open a bash shell with --entrypoint.

**Please check the following, if applicable:**

- [ ] Did you add any tests?
- [x] Did you run `make lint`?
- [] Did you run `make test`? < still running, I'm on a very slow internet connection here. Will comment when done, but I don't expect any problems as the Python codebase hasn't changed.
- [x] Are you making a PR to a feature branch or development rather than master? < development
- [ ] Did you add an entry to `CHANGELOG.rst`? (if not, please do)

metachris commented:
I like the general idea. Will give it a review later today. :+1:

JorritvandenBerg commented:
Thanks! Tests failed due to a problem with an import from neo boa. @localhuman is working on a fix for that.

metachris commented on docker/Dockerfile:
```diff
@@ -56,5 +60,15 @@ RUN pip3 install -e .
-# Example run command
-CMD /bin/bash
+# Remove /neo-python/Chains/SC234 directory at build time 
+RUN rm -rf /neo-python/Chains/SC234/
```
this is not needed, at the beginning the chain is not there

metachris commented on docker/Dockerfile:
```diff
@@ -56,5 +60,15 @@ RUN pip3 install -e .
+ENTRYPOINT ["/entrypoint.sh"]
+
+# Empty command (defaults to running prompt.py without parameters)
+CMD []
```
this is not needed when ENTRYPOINT exists

metachris commented on docker/README.md:
```diff
@@ -29,11 +29,47 @@ See also https://hub.docker.com/r/cityofzion/neo-privatenet
+    $ docker volume rm my-vol
+
+The Docker image supports the following command line parameters:
+-m / --mainnet
```
this is not well formatted

metachris commented on docker/entrypoint.sh:
```diff
@@ -0,0 +1,68 @@
+PROMPT_OPTS=()
+BOOTSTRAP_OPTS=()
+BOOTSTRAP=""
+MAIN_NET=""
```
I feel this file is overly complex. 

 Also some of the vars are not yet used, eg `MAIN_NET`.

metachris commented on docker/entrypoint.sh:
```diff
@@ -0,0 +1,68 @@
+            BOOTSTRAP_OPTS+="-m"
+            MAIN_NET="-m"
+        ;;
+        -p|--privnet)
```
this should throw an error in the bootstrap step, else it downloads the testnet chain

metachris commented on docker/entrypoint.sh:
```diff
@@ -0,0 +1,68 @@
+            PRIV_NET="-p"
+        ;;
+        --coznet)
+            PROMPT_OPTS+="--coznet"
```
this should throw an error in the bootstrap step, else it downloads the testnet chain

metachris commented on docker/entrypoint.sh:
```diff
@@ -0,0 +1,68 @@
+        ;;
+        -c|--config)
+            if [[ "$MAIN_NET" == "-m" ]]; then
+                echo "Can not use config file on main net, proceeding without config file.."
```
uargh, same also goes for other combinations

metachris commented:
First of all, thanks for the effort, it's something I generally like -- making it easier for people to bootstrap the chain.

What I don't like is how complex the `entrypoint.sh` script is, and that it doesn't yet work with a variety of edge-cases and that if has some weird behaviour (eg. downloading the testnet chain when running with `-p`).

I feel there should be a simpler way to accomplish what you want to achieve.

Can you define the target users of this, and the specific use-cases? 

Also, why not have people log in with bash and run bootstrap manually?

If you want to keep working on this, I'm happy to provide more feedback and guidance. We can start merging things into a docker-bootstrap branch and later on merge things into development once it is ready.

metachris commented on docker/Dockerfile:
```diff
@@ -20,7 +20,11 @@
 # Start a container interactively, opening a bash in `/neo-python`, and mounting the current directory as `/neo-python/sc`:
 #
-#    $ docker run --rm -it -v $(pwd):/neo-python/sc --net=host -h neopython --name neopython neopython /bin/bash
+#    $ docker run --rm -it --entrypoint "/bin/bash" -v $(pwd):/neo-python/sc --net=host -h neopython --name neopython neopython
```
I'm not sure about this -- making the default usage more complex. Perhaps it would be better to leave the default entrypoint as is, and use the `--entrypoint` arg when running the bootstrap version,

JorritvandenBerg commented:
Hi Chris,
Thanks for reviewing. I have updated the bash script to fix the edge cases, added error handling and removed the docker ENTRYPOINT. Leaving the choice on wether to run /bin/bash or /entrypoint.sh to the user and not changing the default behaviour. 

The usecase is the ease of bootstrapping and starting prompt.py right away from the Docker CLI / command in a docker-compose.yml. Personally I hardly ever use docker run, let alone running a shell process on a container. I have created a docker-bootstrap branch on my fork, but can not merge this as the branch does not exist on the upstream CoZ repo. Could you create the branch?

localhuman commented:
I'll defer to @metachris on this one

metachris commented:
Just a note -- this PR is not forgotten, I just need a little bit to find some time to play around with a better Docker setup myself. I'll get to it a bit later this week :)

JorritvandenBerg commented:
Thanks, no worries. Let me know if you need any help.

metachris commented:
Just for reference: the Docker compose tooling and setup seems to be focusing here in the future: https://github.com/CityOfZion/neo-local

metachris commented:
Closing this because Docker setup moved to https://github.com/CityOfZion/neo-local. Please consider starting this feature/discussion there 👍
